### ISSUE TITLE
Some RAW vector transfer patterns are not folded away

### ISSUE THREAD

**Issue by hanhanW** [google/iree#8569]
When working on convolution vectorization, I observed that some RAW vector transfer are not folded away because of unit dims. This is needed for properly tiling, fusing and vectorizing fill + convolution. It works in matmul cases, but not convolution cases. Because we intend to set some tiling sizes to 1 for convolution cases. In the context, rank-reduced transfer_read ops are generated during vectorization. This stops the RAW folding pattern.

There is a pattern to fold RAW away: https://github.com/llvm/llvm-project/blob/523c572c376a8998747cf81d16dce8aee69127e5/mlir/lib/Dialect/Vector/IR/VectorOps.cpp#L2916-L2926

> /// 
> ///  %w0 = vector.transfer_write %v0, %arg0[%c1, %c0] {in_bounds = [true, true]}
> ///    : vector<1x4xf32>, tensor<4x4xf32>
> ///  %0 = vector.transfer_read %w0[%c1, %c0], %cf0 {in_bounds = [true, true]}
> ///    : tensor<4x4xf32>, vector<1x4xf32>
> /// 
> ///  -> Folds into
> /// 
> ///  %v0
> /// 

This does not work because the types mismatch in the below example. One is <1x1x4x16xf32>, and another is <1x4x16xf32>. I think we can extend the pattern or rewrite this with `vector.extract_strided_slice`.

```mlir
            %21 = vector.transfer_write %cst, %20[%c0, %c0, %c0, %c0] {in_bounds = [true, true, true, true]} : vector<1x1x4x16xf32>, tensor<1x1x4x16xf32>
            %22 = vector.transfer_read %21[%c0, %c0, %c0, %c0], %cst_0 {in_bounds = [true, true, true]} : tensor<1x1x4x16xf32>, vector<1x4x16xf32>
            %23 = scf.for %arg8 = %c0 to %c3 step %c1 iter_args(%arg9 = %22) -> (vector<1x4x16xf32>) {
```

 Ideally, these ops are eliminated because the source is a constant op. We'll see something like

```mlir
  %23 = scf.for ... iter_args(%arg9 = %cst) -> vector<1x4x16xf32> {
```

@ThomasRaoux do you have any suggestions about this?

**Comment by antiagainst** [comment]
For SPIR-V we invoke [additional patterns](https://github.com/google/iree/blob/main/iree/compiler/Codegen/SPIRV/SPIRVVectorize.cpp#L191-L193
) to cast away the leading one dimensions to allow cancelling read-write ops. You might want to try something similar?

**Comment by hanhanW** [comment]
Ah yes, thanks! I talked to Thomas this morning, and he suggested the same thing. :D

**Comment by pzread** [comment]
Should be fixed in #9089
